### PR TITLE
Allow optionally disabling the static autoloader

### PR DESCRIPTION
**Pull request by legoktm**
The static autoloader presents a problem for environments where pre-PHP
5.6 syntax compatibility is required. Many projects have automated CI
that runs a linter across all files in the repository to prevent
developers from accidentally introdcing incompatible syntax. And while
the static autoloader properly guards itself on lower PHP versions, it
will still trip up linters. Given the number of bugs filed about this,
it appears to be a common setup: #5255, #5316, #5324, and #5407.

This adds a --no-static option to the dump-autoload command, which will
disable the static autoloader by updating autoload_real.php to not look
for the file, and removes autoload_static.php if it exists.

### CODE REVIEW

**Comment by staabm** [comment]
Why does your linter lint external code?

**Comment by legoktm** [comment]
Because regardless of the source of the code, it still has to work on our servers? Doing a generic syntax lint seems to a be a general practice in other projects too, based on the other bugs filed that I linked.

**Comment by Seldaek** [comment]
I am really not so enthusiastic about this. You basically want to add a flag to add support for linting a file you shouldn't lint on PHP versions that are not even receiving security updates anymore. This also adds a bunch of cruft and complexity to a piece of code that's already quite lengthy and convoluted. 

Is it really not possible to exclude that file from the linter config instead? It has quite a specific file name.

**Comment by legoktm** [comment]
Originally we looked into excluding that file from linters (and that was going to be our plan until I saw https://github.com/composer/composer/issues/5651#issuecomment-245191150 which suggested sending a PR), but so far we've found at least 3 different linters in our CI + deployment pipeline that need updating, and some of them are upstream projects (e.g. https://github.com/JakubOnderka/PHP-Parallel-Lint) that don't support excluding specific files (only directories). So we could update the linters, yeah, but after seeing other projects have similar issues (based on the issues filed), I think a solution in composer is warranted.

The code is kind of convoluted, but most of that was to make autoload_real.php look pretty to humans. If you were okay with the indentation of two blocks being wrong, then the code would be cleaner. Or even something like `if (false) {} else { ... }` would be better.

As far as supporting older PHP versions...MediaWiki's current LTS will be supporting 5.5 until 2019 roughly, but that's our choice for our users, and I understand if composer doesn't want to give us first class support.

In conclusion, I see our options as:
1. Fix all the linters
2. Accept this patch
3. Be stuck on composer 1.0.x
4. Make autoload_static.php 5.5 compatible (I kind of assumed someone had researched this and determined it not feasible...is that the case?)
5. ...?

**Comment by paladox** [comment]
+1, and per @legoktm we supported php 5.5 due to the fact we have a lts and will receive updates until, 2019.

But without any configurations we will be stuck on an old composer version until 2019 unless we can approve this patch?

And nope we carnt exclude any files like that so we need something internaly in composer which can and this is what this patch does :)

**Comment by Seldaek** [comment]
How about patching the CI/linter system to allow excluding files? Or just
exclude the whole vendor/composer/ dir if that's an option?

You're telling me changing composer is ok but not changing your other
stuff.. Not sure why composer has to bear the cost here :)

**Comment by legoktm** [comment]
> How about patching the CI/linter system to allow excluding files?

As I stated earlier, that's what we'll have to do if this patch isn't accepted. But I would like to continue this discussion first.

> Or just exclude the whole vendor/composer/ dir if that's an option? 

Excluding vendor/composer doesn't work because we still want to lint vendor/composer/semver.

> You're telling me changing composer is ok but not changing your other stuff.. Not sure why composer has to bear the cost here :)

I submitted a pull request because that was what was suggested in the other ticket. I think the duplicate tickets show that linting everything, including upstream code is a common practice, and should be something that composer supports.

So, what are the actual objections against this? Is it that the code is too convoluted to support? Ideologically against improving support for unsupported PHP versions? Think that linting everything is bad and we should stop? ...

**Comment by alcohol** [comment]
There is no need to lint check your vendor directory. As you said yourself, most projects already do lint checking. On top of that, in fact, one of the linters you linked even recommends explicitly excluding the `vendor` directory.

> https://github.com/JakubOnderka/PHP-Parallel-Lint
>
> Recommended setting for usage with Symfony framework
>
> For run from command line:
```
./bin/parallel-lint --exclude app --exclude vendor .
```

I do not see any justification for moving this into the Composer codebase. Lint checkers are nice, but they are not foolproof (as this case clearly demonstrates). If you choose to build your system on top of this, then dealing with the side effects should be your responsibility imho.

**Comment by Seldaek** [comment]
Another aspect that you would benefit from if you fixed your CI, is that if you disable the static autoloader then essentially you're penalizing users of php 5.6+ which won't get the faster autoloader. Those users that upgrade "early" are usually more performance conscious and as such I don't think that's the right approach.
